nexus/src/app/background/tasks/inventory_load.rs

```diff
@@ -0,0 +1,264 @@
+        tx: watch::Sender<Option<Arc<Collection>>>,
+    ) -> Self {
+        Self { datastore, tx }
+    }
+
+    async fn load_if_needed(&self, opctx: &OpContext) -> InventoryLoadStatus {
+        // Set up a logger for this activation that includes metadata about
+        // the current target.
+        let (old, log) = match &*self.tx.borrow() {
+            None => (None, opctx.log.clone()),
+            Some(old) => {
+                let log = opctx.log.new(slog::o!(
+                    "original_id" => old.id.to_string(),
+                    "original_time_started" => old.time_started.to_string(),
+                ));
+                (Some((old.id, old.time_started)), log)
+            }
+        };
+
+        // Get the ID of the latest collection.
+        let time_loaded = Utc::now();
+        let latest_id = match self
+            .datastore
+            .inventory_get_latest_collection_id(opctx)
+            .await
+        {
+            Ok(Some(id)) => id,
+            Ok(None) => match old {
+                Some((old_id, _)) => {
+                    // We should never go from "some collection" to "no
+                    // collections"; pruning should always keep a small number
+                    // of old collections around until we have new ones to
+                    // replace them.
+                    //
+                    // In this case we won't replace our channel contents with
+                    // `None`; we'll keep around whatever old collection we had
+                    // loaded.
+                    warn!(
+                        log,
+                        "previously had a collection, but now none exist"
+                    );
+                    return InventoryLoadStatus::Error(format!(
+                        "previously loaded collection {old_id}, but now \
+                         no collections exist"
+                    ));
+                }
+                None => {
+                    // Had no collections; still have no collections.
+                    return InventoryLoadStatus::NoCollections;
+                }
+            },
+            Err(err) => {
+                let err = InlineErrorChain::new(&err);
+                warn!(
+                    log,
+                    "failed to read latest inventory collection ID";
+                    &err
+                );
+                return InventoryLoadStatus::Error(format!(
+                    "failed to read latest inventory collection ID: {err}"
+                ));
+            }
+        };
+
+        // Have we already loaded this collection?
+        match old {
+            Some((old_id, old_time_started)) if old_id == latest_id => {
+                debug!(log, "latest inventory collection is unchanged");
+                return InventoryLoadStatus::Loaded {
+                    collection_id: old_id,
+                    time_started: old_time_started,
+                    time_loaded,
+                };
```

Comment:
nitpicky: it could be worth including something in the status message that explicitly indicates whether the collection was newly loaded on this activation or not?

Reply:
I originally had it that way, but I changed it for a maybe-bad reason: I think it makes the omdb expectorate test flaky (in that we have a race between omdb seeing the first activation that loaded a collection vs seeing a subsequent activation that didn't).